pynestml/codegeneration/printers/nest_variable_printer.py

```diff
@@ -0,0 +1,150 @@
+
+from __future__ import annotations
+
+from pynestml.codegeneration.printers.cpp_variable_printer import CppVariablePrinter
+from pynestml.codegeneration.printers.expression_printer import ExpressionPrinter
+from pynestml.codegeneration.nest_unit_converter import NESTUnitConverter
+from pynestml.meta_model.ast_external_variable import ASTExternalVariable
+from pynestml.meta_model.ast_variable import ASTVariable
+from pynestml.symbols.predefined_units import PredefinedUnits
+from pynestml.symbols.predefined_variables import PredefinedVariables
+from pynestml.symbols.symbol import SymbolKind
+from pynestml.symbols.unit_type_symbol import UnitTypeSymbol
+from pynestml.symbols.variable_symbol import BlockType
+from pynestml.utils.ast_utils import ASTUtils
+from pynestml.utils.logger import Logger, LoggingLevel
+from pynestml.utils.messages import Messages
+
+
+class NESTVariablePrinter(CppVariablePrinter):
+    r"""
+    Variable printer for C++ syntax and the NEST API.
+    """
+
+    def __init__(self, expression_printer: ExpressionPrinter, with_origin: bool = True, with_vector_parameter: bool = True) -> None:
+        super().__init__(expression_printer)
+        self.with_origin = with_origin
+        self.with_vector_parameter = with_vector_parameter
+        self._state_symbols = []
+
+    def print_variable(self, variable: ASTVariable) -> str:
+        """
+        Converts a single variable to nest processable format.
+        :param variable: a single variable.
+        :return: a nest processable format.
+        """
+        assert isinstance(variable, ASTVariable)
+
+        if isinstance(variable, ASTExternalVariable):
+            _name = str(variable)
+            if variable.get_alternate_name():
+                # the disadvantage of this approach is that the time the value is to be obtained is not explicitly specified, so we will actually get the value at the end of the min_delay timestep
+                return "((POST_NEURON_TYPE*)(__target))->get_" + variable.get_alternate_name() + "()"
+
+            return "((POST_NEURON_TYPE*)(__target))->get_" + _name + "(_tr_t)"
+
+        if variable.get_name() == PredefinedVariables.E_CONSTANT:
+            return "numerics::e"
+
+        symbol = variable.get_scope().resolve_to_symbol(variable.get_complete_name(), SymbolKind.VARIABLE)
+        if symbol is None:
+            # test if variable name can be resolved to a type
+            if PredefinedUnits.is_unit(variable.get_complete_name()):
+                return str(NESTUnitConverter.get_factor(PredefinedUnits.get_unit(variable.get_complete_name()).get_unit()))
+
+            code, message = Messages.get_could_not_resolve(variable.get_name())
+            Logger.log_message(log_level=LoggingLevel.ERROR, code=code, message=message,
+                               error_position=variable.get_source_position())
+            return ""
+
+        vector_param = ""
+        if self.with_vector_parameter and symbol.has_vector_parameter():
```

Comment:
Why is `with_vector_parameter` necessary?

Reply:
It's because in some cases, a vector variable needs to be printed without the expression in square brackets, where in other cases the size or index parameter in brackets in necessary. An alternative might have been to have a variant of the ``NESTVariablePrinter`` that prints without, and one that prints with this part.